archive/100-theorems-list/inverse-triangle-sum.lean

```diff
@@ -0,0 +1,87 @@
+import tactic
+import algebra.big_operators
+import data.real.basic
+
+lemma range_insert (n : ℕ) :
+insert n (finset.range n) = finset.range (n + 1) :=
+begin
+ext, rw finset.mem_range,
+simp only [finset.mem_insert, finset.mem_range],
+split; omega,
+end
+
+
+lemma discrete_FTC
+{α : Type*}
+[comm_ring α]
+(f F : ℕ → α)
+(hF : F 0 = 0)
+:
+(∀ n,(finset.range n).sum f = F n) ↔
+∀ n, F (n + 1) - F n = f n :=
+begin
+split; intro h,
+
+intro n, rw [← h n, ← h (n+1)],
+rw sub_eq_iff_eq_add,
+rw ← range_insert n, apply finset.sum_insert,
+simp only [finset.not_mem_range_self, not_false_iff],
+
+intro d, induction d with k hk,
+    rw hF, simp only [finset.sum_empty, finset.range_zero],
+rw ← nat.add_one,
+have calc1 := h k,
+rw sub_eq_iff_eq_add at calc1,
+rw calc1, clear calc1,
+rw ← hk,
+rw  ← range_insert,
+simp only [finset.not_mem_range_self, finset.sum_insert, not_false_iff],
+end
+
+-- making this a named function helped me avoid the evils of nat subtraction. It's possible to avoid nat subtraction inline, but it looks kind of arcane / fragile?
+def solution (x : ℚ) : ℚ := if x ≤ 1 then 0 else 1 - 1 / x
+
+lemma inverse_triangle_sum
+(n : ℕ) :
+(finset.range n).sum (λ x, (1 : ℚ) / (x * (x + 1))) = solution (n : ℚ ) :=
```

Comment:
I would avoid making a definition for the RHS of this expression, and just write it as clearly as possible (nat subtractions notwithstanding).